File: lib/py/src/thryft/protocol/_stacked_output_protocol.py
from thryft.protocol._output_protocol import _OutputProtocol


class _StackedOutputProtocol(_OutputProtocol):
    def __init__(self):
        _OutputProtocol.__init__(self)
        self._output_protocol_stack = []

    def write_bool(self, value):
        return self._output_protocol_stack[-1].write_bool(value)

    def write_byte(self, value):
        return self._output_protocol_stack[-1].write_byte(value)

    def write_date_time(self, value):
        return self._output_protocol_stack[-1].write_date_time(value)

    def write_decimal(self, value):
        return self._output_protocol_stack[-1].write_decimal(value)

    def write_email_address(self, value):
        return self._output_protocol_stack[-1].write_email_address(value)

    def write_field_begin(self, name, *args, **kwds):
        return self._output_protocol_stack[-1].write_field_begin(name, *args, **kwds)

    def write_field_end(self):
        return self._output_protocol_stack[-1].write_field_end()

    def write_field_stop(self):
        return self._output_protocol_stack[-1].write_field_stop()

    def write_i16(self, value):
        return self._output_protocol_stack[-1].write_i16(value)

    def write_i32(self, value):
        return self._output_protocol_stack[-1].write_i32(value)

    def write_i64(self, value):
        return self._output_protocol_stack[-1].write_i16(value)

    def write_list_begin(self, *args, **kwds):
        return self._output_protocol_stack[-1].write_list_begin(*args, **kwds)

    def write_list_end(self):
        self._output_protocol_stack.pop(-1)
        return self._output_protocol_stack[-1].write_list_end()

    def write_map_begin(self, *args, **kwds):
        return self._output_protocol_stack[-1].write_map_begin(*args, **kwds)

    def write_map_end(self):
        self._output_protocol_stack.pop(-1)
        return self._output_protocol_stack[-1].write_list_end()

    def write_mixed(self, value):
        return self._output_protocol_stack[-1].write_mixed(value)

    def write_null(self):
        return self._output_protocol_stack[-1].write_null()

    def write_set_begin(self, *args, **kwds):
        return self._output_protocol_stack[-1].write_set_begin(*args, **kwds)

    def write_set_end(self):
        self._output_protocol_stack.pop(-1)
        return self._output_protocol_stack[-1].write_set_end()

    def write_string(self, value):
        return self._output_protocol_stack[-1].write_string(value)

    def write_struct_begin(self, *args, **kwds):
        return self._output_protocol_stack[-1].write_struct_begin(*args, **kwds)

    def write_struct_end(self):
        self._output_protocol_stack.pop(-1)
        return self._output_protocol_stack[-1].write_struct_end()

    def writeUrl(self, value):
        return self._output_protocol_stack[-1].writeUrl(value)

File: lib/java/src/main/java/org/thryft/protocol/StackedOutputProtocol.java
/*******************************************************************************
 * Copyright (c) 2016, Minor Gordon
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 ******************************************************************************/

package org.thryft.protocol;

import java.util.Stack;

public abstract class StackedOutputProtocol<OutputProtocolT extends OutputProtocol>
extends ForwardingOutputProtocol {
    @Override
    public void flush() throws OutputProtocolException {
        if (!outputProtocolStack.isEmpty()) {
            outputProtocolStack.peek().flush();
        }
    }

    @Override
    public void writeListBegin(final Type elementType, final int size)
            throws OutputProtocolException {
        final int protocolStackSize = outputProtocolStack.size();
        outputProtocolStack.peek().writeListBegin(elementType, size);
        if (outputProtocolStack.size() != protocolStackSize + 1) {
            throw new OutputProtocolException(
                    "writeListBegin must add one protocol to the top of the stack");
        }
    }

    @Override
    public void writeListEnd() throws OutputProtocolException {
        outputProtocolStack.pop();
        outputProtocolStack.peek().writeListEnd();
    }

    @Override
    public void writeMapBegin(final Type keyType, final Type valueType,
            final int size) throws OutputProtocolException {
        final int protocolStackSize = outputProtocolStack.size();
        outputProtocolStack.peek().writeMapBegin(keyType, valueType, size);
        if (outputProtocolStack.size() != protocolStackSize + 1) {
            throw new OutputProtocolException(
                    "writeMapBegin must add one protocol to the top of the stack");
        }
    }

    @Override
    public void writeMapEnd() throws OutputProtocolException {
        outputProtocolStack.pop();
        outputProtocolStack.peek().writeMapEnd();
    }

    @Override
    public void writeMessageBegin(final String name, final MessageType type,
            final Object id) throws OutputProtocolException {
        final int protocolStackSize = outputProtocolStack.size();
        outputProtocolStack.peek().writeMessageBegin(name, type, id);
        if (outputProtocolStack.size() != protocolStackSize + 1) {
            throw new OutputProtocolException(
                    "writeMessageBegin must add one protocol to the top of the stack");
        }
    }

    @Override
    public void writeMessageEnd() throws OutputProtocolException {
        outputProtocolStack.pop();
        outputProtocolStack.peek().writeMessageEnd();
    }

    @Override
    public void writeSetBegin(final Type elementType, final int size)
            throws OutputProtocolException {
        outputProtocolStack.peek().writeSetBegin(elementType, size);
    }

    @Override
    public void writeSetEnd() throws OutputProtocolException {
        outputProtocolStack.pop();
        outputProtocolStack.peek().writeSetEnd();
    }

    @Override
    public void writeStructBegin(final String name)
            throws OutputProtocolException {
        final int protocolStackSize = outputProtocolStack.size();
        outputProtocolStack.peek().writeStructBegin(name);
        if (outputProtocolStack.size() != protocolStackSize + 1) {
            throw new OutputProtocolException(
                    "writeStructBegin must add one protocol to the top of the stack");
        }
    }

    @Override
    public void writeStructEnd() throws OutputProtocolException {
        outputProtocolStack.pop();
        outputProtocolStack.peek().writeStructEnd();
    }

    @Override
    public void writeVariant(final Object value) throws OutputProtocolException {
        AbstractOutputProtocol.writeVariant(this, value);
    }

    @Override
    protected OutputProtocol _delegate() {
        return outputProtocolStack.peek();
    }

    protected final Stack<OutputProtocolT> _getOutputProtocolStack() {
        return outputProtocolStack;
    }

    private final Stack<OutputProtocolT> outputProtocolStack = new Stack<OutputProtocolT>();
}
